src/ranch.erl

```diff
@@ -193,24 +193,33 @@ start_error(_, Error) -> Error.
 
 -spec stop_listener(ref()) -> ok | {error, not_found}.
 stop_listener(Ref) ->
+	erpc:call(node(), fun() -> stop_listener1(Ref) end, 5000).
+
+stop_listener1(Ref) ->
+	TransportAndOpts = maybe_get_transport_and_opts(Ref),
+	case supervisor:terminate_child(ranch_sup, {ranch_listener_sup, Ref}) of
+		ok ->
+			_ = supervisor:delete_child(ranch_sup, {ranch_listener_sup, Ref}),
+			ranch_server:cleanup_listener_opts(Ref),
+			stop_listener2(TransportAndOpts);
+		{error, _} = Error ->
+			Error
+	end.
+
+stop_listener2({Transport, TransOpts}) ->
+	Transport:cleanup(TransOpts),
+	ok;
+stop_listener2(undefined) ->
+	ok.
+
+maybe_get_transport_and_opts(Ref) ->
 	try
-		[_, Transport0, _, _, _] = ranch_server:get_listener_start_args(Ref),
-		TransOpts0 = get_transport_options(Ref),
-		{Transport0, TransOpts0}
-	of
-		{Transport, TransOpts} ->
-			case supervisor:terminate_child(ranch_sup, {ranch_listener_sup, Ref}) of
-				ok ->
-					_ = supervisor:delete_child(ranch_sup, {ranch_listener_sup, Ref}),
-					ranch_server:cleanup_listener_opts(Ref),
-					Transport:cleanup(TransOpts),
-					ok;
-				{error, Reason} ->
-					{error, Reason}
-			end
+		[_, Transport, _, _, _] = ranch_server:get_listener_start_args(Ref),
+		TransOpts = get_transport_options(Ref),
+		{Transport, TransOpts}
 	catch
 		error:badarg ->
-			{error, not_found}
+			undefined
```

Comment:
Any reason this was changed?

Reply:
The result of this function is passed on to `stop_listener2`, which already expects a 2-tuple (of `Transports` and `TransOpts`). The idea is to not stop early when the records are not found in `ranch_server` but still call `supervisor:terminate_child` and `supervisor:delete_child` in case the child still exists. If it doesn't, `{error, not_found}` is returned from there. Otherwise if it does, depending on if we got the `Transport` and `TransOpts` here or not, the final step would be to either call `Transport:cleanup(TransOpts)` or do nothing.

Reply:
Right but we clean up the data after stopping not before so I'm not sure how this can happen.

On the other hand I could see a case where the child is terminated but not deleted/cleaned up if the process ends after terminate_child for some reason. In that case we will never clean up even if we stop_listener again.

Perhaps all the functions should be called regardless of what any of them state, and then we look at the end whether the ref/child existed or not.

Reply:
It happens because `ranch_server` monitors the listener process and does a cleanup by itself when it exits, which is what happens when `supervisor:terminate_child` finished. So a cleanup will eventually be done at some time after `terminate_child` has sent the respective message to the supervisor, ie even without the later explicit call to `ranch:cleanup_listener_opts` (not sure why we even have that here, do you know?). If the stop process exits after the terminate message was sent to the supervisor and before the `delete_child` call, to the outside this practically amounts to the `delete_child` call being skipped.

Reply:
Right we may not need this cleanup at all. I'd experiment with only doing terminate+delete then. In that case we'd only return `{error, not_found}` if the delete finds nothing.

Reply:
Done.